### PR TITLE
Preserve backing index ordering for data streams

### PR DESCRIPTION
**Pull request by danhermann**
The resolve index API incorrectly sorts the backing indices returned for data streams. The backing indices on a data stream are an ordered list in which the last index is always the write index so sorting them may produce incorrect results once [migration of existing indices to data streams](https://github.com/elastic/elasticsearch/issues/61046) is permitted.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/admin/indices/resolve/ResolveIndexTests.java [hunk 1]
```diff
@@ -169,6 +169,34 @@ public void testResolveWithMultipleNames() {
         validateDataStreams(dataStreams, "logs-mysql-test");
     }
 
+    public void testResolvePreservesBackingIndexOrdering() {
+        Metadata.Builder builder = Metadata.builder();
+        String dataStreamName = "my-data-stream";
+        List<IndexMetadata> backingIndices = new ArrayList<>();
+        backingIndices.add(createIndexMetadata("not-in-order-2", true));
+        backingIndices.add(createIndexMetadata("not-in-order-1", true));
+        backingIndices.add(createIndexMetadata(DataStream.getDefaultBackingIndexName(dataStreamName, 3), true));
```
Comment (probakowski) [inline server/src/test/java/org/elasticsearch/action/admin/indices/resolve/ResolveIndexTests.java:178]:
```suggestion
        String[] names = new String[]{"not-in-order-2", "not-in-order-1", DataStream.getDefaultBackingIndexName(dataStreamName, 3)};
        List<IndexMetadata> backingIndices = Arrays.stream(names).map(n -> createIndexMetadata(n, true)).collect(Collectors.toList());
```

File: server/src/test/java/org/elasticsearch/action/admin/indices/resolve/ResolveIndexTests.java [hunk 2]
```diff
@@ -169,6 +169,34 @@ public void testResolveWithMultipleNames() {
...
+        List<ResolvedIndex> indices = new ArrayList<>();
+        List<ResolvedAlias> aliases = new ArrayList<>();
+        List<ResolvedDataStream> dataStreams = new ArrayList<>();
+        TransportAction.resolveIndices(new String[]{"*"}, indicesOptions, builder.build(), resolver, indices, aliases, dataStreams, true);
+
+        assertThat(dataStreams.size(), equalTo(1));
+        assertThat(dataStreams.get(0).getBackingIndices().length, equalTo(3));
+        assertThat(dataStreams.get(0).getBackingIndices()[0], equalTo(backingIndices.get(0).getIndex().getName()));
+        assertThat(dataStreams.get(0).getBackingIndices()[1], equalTo(backingIndices.get(1).getIndex().getName()));
+        assertThat(dataStreams.get(0).getBackingIndices()[2], equalTo(backingIndices.get(2).getIndex().getName()));
```
Comment (probakowski) [inline server/src/test/java/org/elasticsearch/action/admin/indices/resolve/ResolveIndexTests.java:197]:
```suggestion
        assertThat(dataStreams.get(0).getBackingIndices(), arrayContaining(names));
```

**Review by probakowski — APPROVED** [review]
LGTM (assuming CI is green), thanks Dan! 
I left suggestion for refactoring in test but it can be ignored.

**Comment by danhermann** [comment]
Thanks, @probakowski! I took your suggestions for simplifying the test.

**Pull request merged**